packages/ra-core/src/routing/useGetPathForRecord.ts

```diff
@@ -49,82 +49,112 @@ export const useGetPathForRecord = <RecordType extends RaRecord = RaRecord>(
-        : resourceDefinition.hasEdit
-          ? 'edit'
-          : false;
-
-    const isLinkFalse =
-        link === false || (link == null && defaultLink === false);
+    const getPathForRecord = useGetRouteForRecordCallback<RecordType>(options);
 
     // we initialize the path with the link value
     const [path, setPath] = useState<string | false | undefined>(() => {
-        if (record == null || isLinkFalse) return false;
-        const linkResult = linkFunc(record, resource) ?? defaultLink;
-        const linkResultIsPromise = isPromise(linkResult);
-        if (linkResultIsPromise) {
-            linkResult.then(resolvedLink => {
-                if (resolvedLink === false) {
-                    // already set to false by default
-                    return;
-                }
-                // update the path when the promise resolves
-                setPath(
-                    createPath({
-                        resource,
-                        id: record.id,
-                        type: resolvedLink,
-                    })
-                );
-            });
-        }
-        return linkResult === false || linkResultIsPromise
-            ? false
-            : createPath({ resource, id: record.id, type: linkResult });
+        getPathForRecord({
+            record,
+            resource,
+            link,
+        }).then(resolvedLink => {
+            if (resolvedLink === false) {
+                // already set to false by default
+                return;
+            }
+            // update the path when the promise resolves
+            setPath(resolvedLink);
+        });
+
+        return false;
     });
 
     // update the path if the record changes
     useEffect(() => {
-        if (record == null || isLinkFalse) {
-            setPath(false);
-            return;
-        }
-        const linkResult = linkFunc(record, resource) ?? defaultLink;
-        const linkResultIsPromise = isPromise(linkResult);
-        if (linkResultIsPromise) {
-            linkResult.then(resolvedLink => {
-                if (resolvedLink === false) {
-                    // already set to false by default
-                    return;
-                }
-                setPath(
-                    createPath({ resource, id: record.id, type: resolvedLink })
-                );
-            });
-        }
-        setPath(
-            linkResult === false || linkResultIsPromise
-                ? false
-                : createPath({ resource, id: record.id, type: linkResult })
-        );
-    }, [createPath, defaultLink, isLinkFalse, linkFunc, record, resource]);
+        getPathForRecord({
+            record,
+            resource,
+            link,
+        }).then(resolvedLink => {
+            // update the path when the promise resolves
+            setPath(resolvedLink);
+        });
+    }, [getPathForRecord, link, record, resource]);
 
     return path;
 };
 
+export const useGetRouteForRecordCallback = <
```

Comment:
I'd call this function useGetRouteForRecordCallback and extract it to another file (and export it)